src/QsCompiler/CompilerOptimizations/Utils.fs

```diff
@@ -0,0 +1,203 @@
+﻿module Microsoft.Quantum.QsCompiler.CompilerOptimization.Utils
+
+open System
+open System.Collections.Generic
+open System.Collections.Immutable
+open System.Numerics
+open Microsoft.Quantum.QsCompiler.DataTypes
+open Microsoft.Quantum.QsCompiler.SyntaxExtensions
+open Microsoft.Quantum.QsCompiler.SyntaxTokens
+open Microsoft.Quantum.QsCompiler.SyntaxTree
+
+
+/// Option default value operator
+let (|?) = defaultArg
+
+
+/// Shorthand for a QsExpressionKind
+type Expr = QsExpressionKind<TypedExpression, Identifier, ResolvedType>
+/// Shorthand for a QsTypeKind
+type TypeKind = QsTypeKind<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation>
+
+
+/// Represents the global dictionary that maps names to callables
+type CallableDict = {
+        compiledCallables: ImmutableDictionary<QsQualifiedName, QsCallable>
+} with
+    member this.getCallable (name: QsQualifiedName): QsCallable =
+        this.compiledCallables.[name]
+
+
+/// Returns whether a given expression is a literal (and thus a constant)
+let rec isLiteral (expr: Expr) (cd: CallableDict): bool =
+    match expr with
+    | UnitValue | IntLiteral _ | BigIntLiteral _ | DoubleLiteral _ | BoolLiteral _ | ResultLiteral _ | PauliLiteral _ -> true
+    | ValueTuple a | StringLiteral (_, a) | ValueArray a -> Seq.forall (fun x -> isLiteral x.Expression cd) a
+    | RangeLiteral (a, b) -> isLiteral a.Expression cd && isLiteral b.Expression cd
+    | NewArray (_, a) -> isLiteral a.Expression cd
+    | Identifier (GlobalCallable _, _) | MissingExpr -> true
+    | CallLikeExpression ({Expression = Identifier (GlobalCallable qualName, _)}, b)
+        when (cd.getCallable qualName).Kind = TypeConstructor -> isLiteral b.Expression cd
+    | CallLikeExpression (a, b) ->
+        isLiteral a.Expression cd && isLiteral b.Expression cd &&
+            TypedExpression.IsPartialApplication (CallLikeExpression (a, b))
+    | _ -> false
+
+
+/// Helper method to improve expression readability
+let rec prettyPrint (expr: Expr): string =
+    match expr with
+    | Identifier (LocalVariable a, _) -> "LocalVariable " + a.Value
+    | Identifier (GlobalCallable a, _) -> "GlobalCallable " + a.Name.Value
+    | CallLikeExpression (f, x) -> "(" + (prettyPrint f.Expression) + " of " + (prettyPrint x.Expression) + ")"
+    | ValueTuple a -> "(" + String.Join(", ", a |> Seq.map (fun x -> x.Expression) |> Seq.map prettyPrint) + ")"
+    | ValueArray a -> "[" + String.Join(", ", a |> Seq.map (fun x -> x.Expression) |> Seq.map prettyPrint) + "]"
+    | ADD (a, b) -> "(" + (prettyPrint a.Expression) + " + " + (prettyPrint b.Expression) + ")"
+    | a -> a.ToString()
+
+
+/// Converts a range literal to a sequence of integers
+let rangeLiteralToSeq (r: Expr): seq<int> =
+    match r with
+    | RangeLiteral (a, b) ->
+        match a.Expression, b.Expression with
+        | IntLiteral start, IntLiteral stop ->
+            seq { int start .. int stop }
```

Comment:
A blind conversion to int is not a good option here (and the same below) - the language does not behave according to specs otherwise. 
You could either define your own sequence generator (this is e.g. what the simulator runtime does: https://github.com/microsoft/qsharp-runtime/blob/master/src/Simulation/Core/Range.cs), or you could refuse to constant fold if a 32-bit int is not enough. I might prefer the former.

Reply:
I fixed this issue in the Loop Unrolling PR. I'll copy the code from that branch to here.